{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0fcb111a-db1c-483d-bfd8-160c46d4201c",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab9f355c-27ac-4635-8b46-7bd01aa0b58b",
   "metadata": {},
   "source": [
    "This notebook creates dummy data for an example GWAS analysis. First, let's load in the needed libraries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "aebfd92e-6f2b-4295-ba8f-a7181a5ec739",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import statsmodels.api as sm\n",
    "from scipy.stats import norm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26420ce4-08fa-47b2-b83b-4a18949c97ae",
   "metadata": {},
   "source": [
    "Next, set chromosome lengths per the hg38 reference assembly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c4f1b4f6-ea59-43ba-ae56-dc1c1050ee3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "chromosome_lengths = {\n",
    "    '1': 248956422, '2': 242193529, '3': 198295559, '4': 190214555, '5': 181538259,\n",
    "    '6': 170805979, '7': 159345973, '8': 145138636, '9': 138394717, '10': 133797422,\n",
    "    '11': 135086622, '12': 133275309, '13': 114364328, '14': 107043718, '15': 101991189,\n",
    "    '16': 90338345, '17': 83257441, '18': 80373285, '19': 58617616, '20': 64444167,\n",
    "    '21': 46709983, '22': 50818468, 'X': 156040895\n",
    "}\n",
    "\n",
    "total_length = sum(chromosome_lengths.values())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6036aa4a-66d4-4109-ae58-84afd4e809bf",
   "metadata": {},
   "source": [
    "Now a function to simulate genotypes based on minor allele frequency."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "838b7e31-16c1-4979-8972-6274e147eb17",
   "metadata": {},
   "outputs": [],
   "source": [
    "def simulate_genotype(maf, n_samples):\n",
    "    return np.random.choice([0, 1, 2], size=n_samples, p=[(1-maf)**2, 2*maf*(1-maf), maf**2])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "873176e8-997c-4d9e-a90c-b6b8ae62788a",
   "metadata": {},
   "source": [
    "Let's set some parameters for our simulated dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5a7ca569-cff0-4737-8628-78de814fc87f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameters\n",
    "n_samples = 50000\n",
    "n_loci = 10000"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d10b3ced-98fd-4d5d-be21-1d366edcc9d2",
   "metadata": {},
   "source": [
    "Now calculate the number of loci per chromosome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "86f19462-485a-43e0-8420-58ee378c5f69",
   "metadata": {},
   "outputs": [],
   "source": [
    "chromosome_loci_counts = {chrom: int(n_loci * length / total_length) for chrom, length in chromosome_lengths.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "598c64ac-a2a9-4f87-9e9e-ee1e267c4197",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1': 821,\n",
       " '2': 799,\n",
       " '3': 654,\n",
       " '4': 627,\n",
       " '5': 598,\n",
       " '6': 563,\n",
       " '7': 525,\n",
       " '8': 478,\n",
       " '9': 456,\n",
       " '10': 441,\n",
       " '11': 445,\n",
       " '12': 439,\n",
       " '13': 377,\n",
       " '14': 353,\n",
       " '15': 336,\n",
       " '16': 298,\n",
       " '17': 274,\n",
       " '18': 265,\n",
       " '19': 193,\n",
       " '20': 212,\n",
       " '21': 154,\n",
       " '22': 167,\n",
       " 'X': 514}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chromosome_loci_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6ccf2fe-14e5-41d0-a88d-2e5b9246ee30",
   "metadata": {},
   "source": [
    "Create the info columns for the PED file. First, let's create a balanced set of phenotypes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "17fa2b9f-d54e-4e81-8b54-9632565c1d3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotypes = np.array([0] * (n_samples // 2) + [1] * (n_samples // 2))\n",
    "np.random.shuffle(phenotypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ff0c857f-fa82-4c3e-b416-0fb9862c6fb6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([0, 1]), array([25000, 25000]))"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(phenotypes, return_counts=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "844542d2-3495-4a09-b363-f66591756ac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "PED_info_columns = pd.DataFrame({\n",
    "    'FID': [0] * n_samples,\n",
    "    'IID': [f'IND{i}' for i in range(n_samples)],\n",
    "    'Father_ID': [0] * n_samples,\n",
    "    'Mother_ID': [0] * n_samples,\n",
    "    'Sex': [0] * n_samples,\n",
    "    'PHENOTYPE': phenotypes\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4675da4e-bd35-4396-9176-c0e146003b26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FID</th>\n",
       "      <th>IID</th>\n",
       "      <th>Father_ID</th>\n",
       "      <th>Mother_ID</th>\n",
       "      <th>Sex</th>\n",
       "      <th>PHENOTYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>IND0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>IND1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>IND2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>IND3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>IND4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   FID   IID  Father_ID  Mother_ID  Sex  PHENOTYPE\n",
       "0    0  IND0          0          0    0          1\n",
       "1    0  IND1          0          0    0          1\n",
       "2    0  IND2          0          0    0          0\n",
       "3    0  IND3          0          0    0          0\n",
       "4    0  IND4          0          0    0          0"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PED_info_columns.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23a41a0c-0136-4ada-a333-5690e4e645ef",
   "metadata": {},
   "source": [
    "Create lists to hold the generated data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "22034309-c6b3-4236-ad57-8800941fc4a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "genotype_columns = []\n",
    "variant_info = []"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ac0d293-0f49-4d6a-9536-4cd3375341c8",
   "metadata": {},
   "source": [
    "Now to generate genotypes per variant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e06e071f-9a1e-48ef-9595-0be6a1ce88c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "for chrom, n_variants in chromosome_loci_counts.items():\n",
    "    chrom_length = chromosome_lengths.get(chrom)\n",
    "    \n",
    "    for i in range(n_variants):\n",
    "        snp_id = f\"rs{chrom}_{i+1}\"\n",
    "        maf = np.random.uniform(0.01, 0.5)  # Random MAF between 0.01 and 0.5\n",
    "        genotypes = simulate_genotype(maf, n_samples)\n",
    "        \n",
    "        # Append genotype data\n",
    "        genotype_columns.append(pd.Series(genotypes, name=snp_id))\n",
    "\n",
    "        # Variant position\n",
    "        position = np.random.randint(1, chrom_length + 1)\n",
    "        \n",
    "        # Store variant information\n",
    "        variant_info.append([chrom, snp_id, 0, position])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6ba62d14-b9ce-4d72-be89-13ed52900e94",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "genotype_data = pd.concat(genotype_columns, axis = 1)\n",
    "PED = pd.concat([PED_info_columns, genotype_data], axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "cb516e76-55be-4215-8cff-0a0e9effe389",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FID</th>\n",
       "      <th>IID</th>\n",
       "      <th>Father_ID</th>\n",
       "      <th>Mother_ID</th>\n",
       "      <th>Sex</th>\n",
       "      <th>PHENOTYPE</th>\n",
       "      <th>rs1_1</th>\n",
       "      <th>rs1_2</th>\n",
       "      <th>rs1_3</th>\n",
       "      <th>rs1_4</th>\n",
       "      <th>...</th>\n",
       "      <th>rsX_505</th>\n",
       "      <th>rsX_506</th>\n",
       "      <th>rsX_507</th>\n",
       "      <th>rsX_508</th>\n",
       "      <th>rsX_509</th>\n",
       "      <th>rsX_510</th>\n",
       "      <th>rsX_511</th>\n",
       "      <th>rsX_512</th>\n",
       "      <th>rsX_513</th>\n",
       "      <th>rsX_514</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>IND0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>IND1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>IND2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>IND3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>IND4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0</td>\n",
       "      <td>IND5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0</td>\n",
       "      <td>IND6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0</td>\n",
       "      <td>IND7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0</td>\n",
       "      <td>IND8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0</td>\n",
       "      <td>IND9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 9995 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   FID   IID  Father_ID  Mother_ID  Sex  PHENOTYPE  rs1_1  rs1_2  rs1_3  \\\n",
       "0    0  IND0          0          0    0          1      1      0      0   \n",
       "1    0  IND1          0          0    0          1      0      0      0   \n",
       "2    0  IND2          0          0    0          0      1      0      0   \n",
       "3    0  IND3          0          0    0          0      1      0      0   \n",
       "4    0  IND4          0          0    0          0      2      0      1   \n",
       "5    0  IND5          0          0    0          1      1      0      1   \n",
       "6    0  IND6          0          0    0          1      0      1      2   \n",
       "7    0  IND7          0          0    0          0      2      0      1   \n",
       "8    0  IND8          0          0    0          1      0      0      2   \n",
       "9    0  IND9          0          0    0          1      0      0      1   \n",
       "\n",
       "   rs1_4  ...  rsX_505  rsX_506  rsX_507  rsX_508  rsX_509  rsX_510  rsX_511  \\\n",
       "0      0  ...        2        0        1        0        1        0        0   \n",
       "1      0  ...        1        1        0        0        0        1        0   \n",
       "2      0  ...        0        0        1        0        0        0        0   \n",
       "3      0  ...        0        0        1        0        0        1        1   \n",
       "4      0  ...        0        1        0        0        0        0        0   \n",
       "5      0  ...        1        0        0        0        0        1        0   \n",
       "6      0  ...        0        2        1        0        0        0        0   \n",
       "7      0  ...        0        0        0        0        0        0        1   \n",
       "8      0  ...        0        0        0        0        0        0        0   \n",
       "9      0  ...        0        0        1        0        0        0        2   \n",
       "\n",
       "   rsX_512  rsX_513  rsX_514  \n",
       "0        1        0        0  \n",
       "1        0        0        0  \n",
       "2        1        1        0  \n",
       "3        0        0        0  \n",
       "4        2        2        0  \n",
       "5        0        0        0  \n",
       "6        2        0        0  \n",
       "7        1        1        0  \n",
       "8        1        1        1  \n",
       "9        2        0        2  \n",
       "\n",
       "[10 rows x 9995 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PED.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "34a73df8-9186-4f38-8e68-ce594b1d1cac",
   "metadata": {},
   "outputs": [],
   "source": [
    "MAP = pd.DataFrame(variant_info, columns = ['chr','rsID','cM','pos'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "65b424a1-71c4-4e3b-9546-315b8ff2abc1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>rsID</th>\n",
       "      <th>cM</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>rs1_1</td>\n",
       "      <td>0</td>\n",
       "      <td>200826415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>rs1_2</td>\n",
       "      <td>0</td>\n",
       "      <td>4027976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>rs1_3</td>\n",
       "      <td>0</td>\n",
       "      <td>222620360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>rs1_4</td>\n",
       "      <td>0</td>\n",
       "      <td>5556784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>rs1_5</td>\n",
       "      <td>0</td>\n",
       "      <td>176260783</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr   rsID  cM        pos\n",
       "0   1  rs1_1   0  200826415\n",
       "1   1  rs1_2   0    4027976\n",
       "2   1  rs1_3   0  222620360\n",
       "3   1  rs1_4   0    5556784\n",
       "4   1  rs1_5   0  176260783"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MAP.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "cd835fa5-01c9-4e0d-8eaf-d0cfe3a267a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_GWAS_summary_stats_from_PED(PED_file):\n",
    "    # Read PED file\n",
    "    df = PED_file\n",
    " \n",
    "    # Extract phenotype and genotypes\n",
    "    phenotypes = df.iloc[:, 5].astype(int)\n",
    "    genotypes = df.iloc[:, 6:].astype(int)\n",
    "    \n",
    "    # Initialize lists to store results\n",
    "    snp_ids = genotypes.columns.tolist()\n",
    "    allele_freqs = []\n",
    "    betas = []\n",
    "    odds_ratios = []\n",
    "    p_values = []\n",
    "\n",
    "    # Loop through each SNP\n",
    "    for snp in snp_ids:\n",
    "        snp_genotypes = genotypes[snp]\n",
    "\n",
    "        # Calculate allele frequencies\n",
    "        allele_freq = snp_genotypes.mean() / 2\n",
    "        allele_freqs.append(allele_freq)\n",
    "        \n",
    "        # Logistic regression for association testing\n",
    "        X = sm.add_constant(snp_genotypes)\n",
    "        model = sm.Logit(phenotypes, X)\n",
    "        result = model.fit(disp=0)\n",
    "        \n",
    "        # Get beta, odds ratio, and p-value\n",
    "        beta = result.params[snp]\n",
    "        odds_ratio = np.exp(beta)\n",
    "        p_value = result.pvalues[snp]\n",
    "        \n",
    "        betas.append(beta)\n",
    "        odds_ratios.append(odds_ratio)\n",
    "        p_values.append(p_value)\n",
    "\n",
    "    # Create summary statistics dataframe\n",
    "    summary_stats = pd.DataFrame({\n",
    "        'SNP': snp_ids,\n",
    "        'Allele_Frequency': allele_freqs,\n",
    "        'Beta': betas,\n",
    "        'Odds_Ratio': odds_ratios,\n",
    "        'P_Value': p_values\n",
    "    })\n",
    "\n",
    "    return summary_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "710dfaff-f29a-431b-86f9-d26b21a61dc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_stats = calculate_GWAS_summary_stats_from_PED(PED)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1c364cb5-c11b-4e23-937c-88bc2a79c5d7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>Allele_Frequency</th>\n",
       "      <th>Beta</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>P_Value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1_1</td>\n",
       "      <td>0.42803</td>\n",
       "      <td>0.007762</td>\n",
       "      <td>1.007793</td>\n",
       "      <td>0.543708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1_2</td>\n",
       "      <td>0.06738</td>\n",
       "      <td>0.002549</td>\n",
       "      <td>1.002552</td>\n",
       "      <td>0.919577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1_3</td>\n",
       "      <td>0.18924</td>\n",
       "      <td>-0.008810</td>\n",
       "      <td>0.991229</td>\n",
       "      <td>0.584180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1_4</td>\n",
       "      <td>0.05474</td>\n",
       "      <td>-0.024823</td>\n",
       "      <td>0.975482</td>\n",
       "      <td>0.372813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1_5</td>\n",
       "      <td>0.12379</td>\n",
       "      <td>-0.013784</td>\n",
       "      <td>0.986310</td>\n",
       "      <td>0.472171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs1_6</td>\n",
       "      <td>0.22503</td>\n",
       "      <td>-0.001488</td>\n",
       "      <td>0.998513</td>\n",
       "      <td>0.921653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rs1_7</td>\n",
       "      <td>0.40321</td>\n",
       "      <td>-0.004071</td>\n",
       "      <td>0.995938</td>\n",
       "      <td>0.752156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>rs1_8</td>\n",
       "      <td>0.25622</td>\n",
       "      <td>-0.018303</td>\n",
       "      <td>0.981864</td>\n",
       "      <td>0.207007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>rs1_9</td>\n",
       "      <td>0.32605</td>\n",
       "      <td>-0.014779</td>\n",
       "      <td>0.985330</td>\n",
       "      <td>0.272441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>rs1_10</td>\n",
       "      <td>0.40165</td>\n",
       "      <td>0.006437</td>\n",
       "      <td>1.006458</td>\n",
       "      <td>0.618614</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      SNP  Allele_Frequency      Beta  Odds_Ratio   P_Value\n",
       "0   rs1_1           0.42803  0.007762    1.007793  0.543708\n",
       "1   rs1_2           0.06738  0.002549    1.002552  0.919577\n",
       "2   rs1_3           0.18924 -0.008810    0.991229  0.584180\n",
       "3   rs1_4           0.05474 -0.024823    0.975482  0.372813\n",
       "4   rs1_5           0.12379 -0.013784    0.986310  0.472171\n",
       "5   rs1_6           0.22503 -0.001488    0.998513  0.921653\n",
       "6   rs1_7           0.40321 -0.004071    0.995938  0.752156\n",
       "7   rs1_8           0.25622 -0.018303    0.981864  0.207007\n",
       "8   rs1_9           0.32605 -0.014779    0.985330  0.272441\n",
       "9  rs1_10           0.40165  0.006437    1.006458  0.618614"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_stats.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a0237088-ab98-4769-89e4-e300bce8daae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>Allele_Frequency</th>\n",
       "      <th>Beta</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>P_Value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>469</th>\n",
       "      <td>rs1_470</td>\n",
       "      <td>0.28793</td>\n",
       "      <td>0.047525</td>\n",
       "      <td>1.048673</td>\n",
       "      <td>0.000671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1169</th>\n",
       "      <td>rs2_349</td>\n",
       "      <td>0.45021</td>\n",
       "      <td>0.046613</td>\n",
       "      <td>1.047717</td>\n",
       "      <td>0.000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1836</th>\n",
       "      <td>rs3_217</td>\n",
       "      <td>0.42116</td>\n",
       "      <td>-0.048621</td>\n",
       "      <td>0.952542</td>\n",
       "      <td>0.000141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2656</th>\n",
       "      <td>rs4_383</td>\n",
       "      <td>0.24153</td>\n",
       "      <td>0.048885</td>\n",
       "      <td>1.050099</td>\n",
       "      <td>0.000976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2711</th>\n",
       "      <td>rs4_438</td>\n",
       "      <td>0.34079</td>\n",
       "      <td>-0.045140</td>\n",
       "      <td>0.955864</td>\n",
       "      <td>0.000719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2979</th>\n",
       "      <td>rs5_79</td>\n",
       "      <td>0.12279</td>\n",
       "      <td>0.067004</td>\n",
       "      <td>1.069300</td>\n",
       "      <td>0.000508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3147</th>\n",
       "      <td>rs5_247</td>\n",
       "      <td>0.47287</td>\n",
       "      <td>0.044800</td>\n",
       "      <td>1.045818</td>\n",
       "      <td>0.000403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3200</th>\n",
       "      <td>rs5_300</td>\n",
       "      <td>0.26862</td>\n",
       "      <td>0.052631</td>\n",
       "      <td>1.054041</td>\n",
       "      <td>0.000230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3577</th>\n",
       "      <td>rs6_79</td>\n",
       "      <td>0.29646</td>\n",
       "      <td>0.051740</td>\n",
       "      <td>1.053102</td>\n",
       "      <td>0.000186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6828</th>\n",
       "      <td>rs12_422</td>\n",
       "      <td>0.13725</td>\n",
       "      <td>-0.061698</td>\n",
       "      <td>0.940166</td>\n",
       "      <td>0.000795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7044</th>\n",
       "      <td>rs13_199</td>\n",
       "      <td>0.17755</td>\n",
       "      <td>0.067266</td>\n",
       "      <td>1.069580</td>\n",
       "      <td>0.000049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8329</th>\n",
       "      <td>rs17_120</td>\n",
       "      <td>0.13030</td>\n",
       "      <td>0.063303</td>\n",
       "      <td>1.065350</td>\n",
       "      <td>0.000765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8366</th>\n",
       "      <td>rs17_157</td>\n",
       "      <td>0.20158</td>\n",
       "      <td>-0.056865</td>\n",
       "      <td>0.944722</td>\n",
       "      <td>0.000300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9242</th>\n",
       "      <td>rs21_89</td>\n",
       "      <td>0.45937</td>\n",
       "      <td>-0.050350</td>\n",
       "      <td>0.950896</td>\n",
       "      <td>0.000071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9282</th>\n",
       "      <td>rs21_129</td>\n",
       "      <td>0.28565</td>\n",
       "      <td>0.051673</td>\n",
       "      <td>1.053031</td>\n",
       "      <td>0.000202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9480</th>\n",
       "      <td>rsX_6</td>\n",
       "      <td>0.16634</td>\n",
       "      <td>-0.062030</td>\n",
       "      <td>0.939855</td>\n",
       "      <td>0.000270</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNP  Allele_Frequency      Beta  Odds_Ratio   P_Value\n",
       "469    rs1_470           0.28793  0.047525    1.048673  0.000671\n",
       "1169   rs2_349           0.45021  0.046613    1.047717  0.000240\n",
       "1836   rs3_217           0.42116 -0.048621    0.952542  0.000141\n",
       "2656   rs4_383           0.24153  0.048885    1.050099  0.000976\n",
       "2711   rs4_438           0.34079 -0.045140    0.955864  0.000719\n",
       "2979    rs5_79           0.12279  0.067004    1.069300  0.000508\n",
       "3147   rs5_247           0.47287  0.044800    1.045818  0.000403\n",
       "3200   rs5_300           0.26862  0.052631    1.054041  0.000230\n",
       "3577    rs6_79           0.29646  0.051740    1.053102  0.000186\n",
       "6828  rs12_422           0.13725 -0.061698    0.940166  0.000795\n",
       "7044  rs13_199           0.17755  0.067266    1.069580  0.000049\n",
       "8329  rs17_120           0.13030  0.063303    1.065350  0.000765\n",
       "8366  rs17_157           0.20158 -0.056865    0.944722  0.000300\n",
       "9242   rs21_89           0.45937 -0.050350    0.950896  0.000071\n",
       "9282  rs21_129           0.28565  0.051673    1.053031  0.000202\n",
       "9480     rsX_6           0.16634 -0.062030    0.939855  0.000270"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_stats[summary_stats['P_Value'] < 0.001]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "794be7fb-7920-470c-85c2-f18e4de9c1c9",
   "metadata": {},
   "source": [
    "There are likely a few loci that approach significance but none that reach a genome-wide threshold. Let's manually add a few. We will write a function that creates a genotype distribution based on specified genotype frequencies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5a349fdf-faba-43e8-b8a0-fc95c4acefba",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_genotypes_for_significant_gwas(phenotypes, controls_genotype_frequencies, cases_genotype_frequencies):\n",
    "    n_samples = len(phenotypes)\n",
    "    n_cases = np.sum(phenotypes)\n",
    "    n_controls = n_samples - n_cases\n",
    "    \n",
    "    # Generate control genotypes\n",
    "    control_genotypes = np.random.choice(\n",
    "        [0, 1, 2], size=n_controls, p=controls_genotype_frequencies)\n",
    "    \n",
    "    # Generate case genotypes\n",
    "    case_genotypes = np.random.choice(\n",
    "        [0, 1, 2], size=n_cases, p=cases_genotype_frequencies)\n",
    "    \n",
    "    # Combine genotypes\n",
    "    genotypes = np.zeros(n_samples, dtype=int)\n",
    "    genotypes[phenotypes == 0] = control_genotypes\n",
    "    genotypes[phenotypes == 1] = case_genotypes\n",
    "    \n",
    "    return genotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "775d4bd1-a4a0-4930-84c3-2fd0efdd04a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_variant_1 = generate_genotypes_for_significant_gwas(PED['PHENOTYPE'], [0.40, 0.38, 0.22], [0.35, 0.42, 0.23])\n",
    "sig_variant_2 = generate_genotypes_for_significant_gwas(PED['PHENOTYPE'], [0.40, 0.38, 0.22], [0.35, 0.42, 0.23])\n",
    "sig_variant_3 = generate_genotypes_for_significant_gwas(PED['PHENOTYPE'], [0.40, 0.38, 0.22], [0.35, 0.42, 0.23])\n",
    "sig_variant_4 = generate_genotypes_for_significant_gwas(PED['PHENOTYPE'], [0.40, 0.38, 0.22], [0.35, 0.42, 0.23])\n",
    "sig_variant_5 = generate_genotypes_for_significant_gwas(PED['PHENOTYPE'], [0.40, 0.37, 0.23], [0.35, 0.42, 0.23])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5d55e2c-f586-407b-8499-efa7e0c83763",
   "metadata": {},
   "source": [
    "Add these variants to the PED and MAP files. We'll add variants to chromosomes 2, 3, 4, 16, and 17. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "1f1c017d-f817-427d-a2c4-536e0e47a096",
   "metadata": {},
   "outputs": [],
   "source": [
    "PED['rs848293'] = sig_variant_1\n",
    "PED['rs10779987'] = sig_variant_2\n",
    "PED['rs62340585'] = sig_variant_3\n",
    "PED['rs11865086'] = sig_variant_4\n",
    "PED['rs4792891'] = sig_variant_5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "986864c6-a4d6-4baf-8d88-730d8f839dbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_MAP_rows = pd.DataFrame({\n",
    "    'chr': [2,3,4,16,17],\n",
    "    'rsID': ['rs848293','rs10779987','rs62340585','rs11865086','rs4792891'],\n",
    "    'cM': [0,0,0,0,0],\n",
    "    'pos': [58155355,81921705,175949971,30119172,45896132]\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "2db10476-9779-41f7-a97e-26ff2c068b8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "MAP = pd.concat([MAP, new_MAP_rows], ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "e938bf04-8618-4936-8259-223aa650b490",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>rsID</th>\n",
       "      <th>cM</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9989</th>\n",
       "      <td>2</td>\n",
       "      <td>rs848293</td>\n",
       "      <td>0</td>\n",
       "      <td>58155355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9990</th>\n",
       "      <td>3</td>\n",
       "      <td>rs10779987</td>\n",
       "      <td>0</td>\n",
       "      <td>81921705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9991</th>\n",
       "      <td>4</td>\n",
       "      <td>rs62340585</td>\n",
       "      <td>0</td>\n",
       "      <td>175949971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9992</th>\n",
       "      <td>16</td>\n",
       "      <td>rs11865086</td>\n",
       "      <td>0</td>\n",
       "      <td>30119172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9993</th>\n",
       "      <td>17</td>\n",
       "      <td>rs4792891</td>\n",
       "      <td>0</td>\n",
       "      <td>45896132</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr        rsID  cM        pos\n",
       "9989   2    rs848293   0   58155355\n",
       "9990   3  rs10779987   0   81921705\n",
       "9991   4  rs62340585   0  175949971\n",
       "9992  16  rs11865086   0   30119172\n",
       "9993  17   rs4792891   0   45896132"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MAP.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d308ee0c-297c-45fa-b862-38abfde0a3d0",
   "metadata": {},
   "source": [
    "Now, recalculate the summary statistics. The results should include our very significant loci."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "9741d38e-a414-4fc0-9a8d-592ed9d2ab52",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_stats = calculate_GWAS_summary_stats_from_PED(PED)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e2260ffd-0fa8-4f6c-ae47-6147d2c58432",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>Allele_Frequency</th>\n",
       "      <th>Beta</th>\n",
       "      <th>Odds_Ratio</th>\n",
       "      <th>P_Value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>469</th>\n",
       "      <td>rs1_470</td>\n",
       "      <td>0.28793</td>\n",
       "      <td>0.047525</td>\n",
       "      <td>1.048673</td>\n",
       "      <td>6.709569e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1169</th>\n",
       "      <td>rs2_349</td>\n",
       "      <td>0.45021</td>\n",
       "      <td>0.046613</td>\n",
       "      <td>1.047717</td>\n",
       "      <td>2.398810e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1836</th>\n",
       "      <td>rs3_217</td>\n",
       "      <td>0.42116</td>\n",
       "      <td>-0.048621</td>\n",
       "      <td>0.952542</td>\n",
       "      <td>1.413980e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2656</th>\n",
       "      <td>rs4_383</td>\n",
       "      <td>0.24153</td>\n",
       "      <td>0.048885</td>\n",
       "      <td>1.050099</td>\n",
       "      <td>9.759826e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2711</th>\n",
       "      <td>rs4_438</td>\n",
       "      <td>0.34079</td>\n",
       "      <td>-0.045140</td>\n",
       "      <td>0.955864</td>\n",
       "      <td>7.191784e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2979</th>\n",
       "      <td>rs5_79</td>\n",
       "      <td>0.12279</td>\n",
       "      <td>0.067004</td>\n",
       "      <td>1.069300</td>\n",
       "      <td>5.082493e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3147</th>\n",
       "      <td>rs5_247</td>\n",
       "      <td>0.47287</td>\n",
       "      <td>0.044800</td>\n",
       "      <td>1.045818</td>\n",
       "      <td>4.032052e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3200</th>\n",
       "      <td>rs5_300</td>\n",
       "      <td>0.26862</td>\n",
       "      <td>0.052631</td>\n",
       "      <td>1.054041</td>\n",
       "      <td>2.295116e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3577</th>\n",
       "      <td>rs6_79</td>\n",
       "      <td>0.29646</td>\n",
       "      <td>0.051740</td>\n",
       "      <td>1.053102</td>\n",
       "      <td>1.863902e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6828</th>\n",
       "      <td>rs12_422</td>\n",
       "      <td>0.13725</td>\n",
       "      <td>-0.061698</td>\n",
       "      <td>0.940166</td>\n",
       "      <td>7.950125e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7044</th>\n",
       "      <td>rs13_199</td>\n",
       "      <td>0.17755</td>\n",
       "      <td>0.067266</td>\n",
       "      <td>1.069580</td>\n",
       "      <td>4.861334e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8329</th>\n",
       "      <td>rs17_120</td>\n",
       "      <td>0.13030</td>\n",
       "      <td>0.063303</td>\n",
       "      <td>1.065350</td>\n",
       "      <td>7.652113e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8366</th>\n",
       "      <td>rs17_157</td>\n",
       "      <td>0.20158</td>\n",
       "      <td>-0.056865</td>\n",
       "      <td>0.944722</td>\n",
       "      <td>2.997738e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9242</th>\n",
       "      <td>rs21_89</td>\n",
       "      <td>0.45937</td>\n",
       "      <td>-0.050350</td>\n",
       "      <td>0.950896</td>\n",
       "      <td>7.122184e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9282</th>\n",
       "      <td>rs21_129</td>\n",
       "      <td>0.28565</td>\n",
       "      <td>0.051673</td>\n",
       "      <td>1.053031</td>\n",
       "      <td>2.015816e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9480</th>\n",
       "      <td>rsX_6</td>\n",
       "      <td>0.16634</td>\n",
       "      <td>-0.062030</td>\n",
       "      <td>0.939855</td>\n",
       "      <td>2.702682e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9989</th>\n",
       "      <td>rs848293</td>\n",
       "      <td>0.42448</td>\n",
       "      <td>0.109197</td>\n",
       "      <td>1.115382</td>\n",
       "      <td>1.539571e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9990</th>\n",
       "      <td>rs10779987</td>\n",
       "      <td>0.42523</td>\n",
       "      <td>0.123588</td>\n",
       "      <td>1.131550</td>\n",
       "      <td>1.100377e-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9991</th>\n",
       "      <td>rs62340585</td>\n",
       "      <td>0.42429</td>\n",
       "      <td>0.099752</td>\n",
       "      <td>1.104897</td>\n",
       "      <td>2.099997e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9992</th>\n",
       "      <td>rs11865086</td>\n",
       "      <td>0.42464</td>\n",
       "      <td>0.098775</td>\n",
       "      <td>1.103818</td>\n",
       "      <td>4.821300e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9993</th>\n",
       "      <td>rs4792891</td>\n",
       "      <td>0.43093</td>\n",
       "      <td>0.060634</td>\n",
       "      <td>1.062509</td>\n",
       "      <td>2.240537e-07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             SNP  Allele_Frequency      Beta  Odds_Ratio       P_Value\n",
       "469      rs1_470           0.28793  0.047525    1.048673  6.709569e-04\n",
       "1169     rs2_349           0.45021  0.046613    1.047717  2.398810e-04\n",
       "1836     rs3_217           0.42116 -0.048621    0.952542  1.413980e-04\n",
       "2656     rs4_383           0.24153  0.048885    1.050099  9.759826e-04\n",
       "2711     rs4_438           0.34079 -0.045140    0.955864  7.191784e-04\n",
       "2979      rs5_79           0.12279  0.067004    1.069300  5.082493e-04\n",
       "3147     rs5_247           0.47287  0.044800    1.045818  4.032052e-04\n",
       "3200     rs5_300           0.26862  0.052631    1.054041  2.295116e-04\n",
       "3577      rs6_79           0.29646  0.051740    1.053102  1.863902e-04\n",
       "6828    rs12_422           0.13725 -0.061698    0.940166  7.950125e-04\n",
       "7044    rs13_199           0.17755  0.067266    1.069580  4.861334e-05\n",
       "8329    rs17_120           0.13030  0.063303    1.065350  7.652113e-04\n",
       "8366    rs17_157           0.20158 -0.056865    0.944722  2.997738e-04\n",
       "9242     rs21_89           0.45937 -0.050350    0.950896  7.122184e-05\n",
       "9282    rs21_129           0.28565  0.051673    1.053031  2.015816e-04\n",
       "9480       rsX_6           0.16634 -0.062030    0.939855  2.702682e-04\n",
       "9989    rs848293           0.42448  0.109197    1.115382  1.539571e-20\n",
       "9990  rs10779987           0.42523  0.123588    1.131550  1.100377e-25\n",
       "9991  rs62340585           0.42429  0.099752    1.104897  2.099997e-17\n",
       "9992  rs11865086           0.42464  0.098775    1.103818  4.821300e-17\n",
       "9993   rs4792891           0.43093  0.060634    1.062509  2.240537e-07"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_stats[summary_stats['P_Value'] < 0.001]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74598f46-2c88-4b38-a5db-6b400c8d503b",
   "metadata": {},
   "source": [
    "Save the dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "730521fb-cea0-4ef7-bdec-93726e56a6fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "MAP.to_csv('GWAS_tutorial.map', sep = '\\t', header = True, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "2e022f86-f483-44bb-b6d8-4af9919d3c83",
   "metadata": {},
   "outputs": [],
   "source": [
    "PED.to_csv('GWAS_tutorial.ped', sep = '\\t', header = True, index = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
